[JS/TS] PojoDefinedByConsArgs attribute

alfonsogarciacaro opened the pull request:
See https://github.com/fable-compiler/Fable/issues/4025#issuecomment-2636526013

This adapts @MangelMaxime work in #4007 to use a single attribute instead of the combination `Global/ParamObject/Emit`. This simplifies the declaration and allows for the interface to be exported in TS as export rules don't conflict with `Global` attribute.

The name `PojoDefinedByConsArgs` is tentative, please let me know if you prefer to change it.

It's a bit of a tricky feature, as many things can go wrong if the user doesn't declare the type correctly, but let's see how it goes :)

MangelMaxime commented:
`PojoDefinedByConsArgs` seems like a complex attribute name even if it is quite describe of the internals.

What about using something like:

- `Pojo` as this is what we are trying to generate (I think in Fable 1 or 2 we had a `Pojo` that we removed but because it was a long time ago I think this should be ok)
- `PojoInterface` if we want to avoid re-using an old attribute and have one which seems less intimidating.

If I understand it also means that the old `Global`/`ParamObject` pattern can be fully replaced by the new attribute right? Perhaps, we could generate a warning to ask the user to migrate to the new attribute?

alfonsogarciacaro commented:
`PojoDefinedByConsArgs` is definitely ugly 😅 `Pojo` could work, I was thinking that we did have something like but I don't remember how it worked.

> Perhaps, we could generate a warning to ask the user to migrate to the new attribute?

ParamObject + Emit will still work, but it is indeed a good idea to emit a warning to encourage the use of the new attribute ...I just hope users don't start decorating everything with Pojo and ask it to support everything (reflection, etc) as it often happens :)

MangelMaxime commented:
> `PojoDefinedByConsArgs` is definitely ugly 😅 `Pojo` could work, I was thinking that we did have something like but I don't remember how it worked.

If I remember correctly, we could decorated a type with `[<Pojo>]` to emit a Pojo instead of a standard class or something like that. But with the move to Fable 2 or 3, we changed the representation of the some types and they ended up compatible with code expecting Pojo so we were able to remove that attribute. Probably thanks to the inheritance to `Record` class.

> I just hope users don't start decorating everything with Pojo and ask it to support everything (reflection, etc) as it often happens :)

I hope not too.

I think `[<Pojo>]` should be limited to interop code and if someone want to use reflection they should use normal interface/classes because native code will not be able to consume the reflection code (at least not easily).

Also, interfaces are still more powerful in certain usage because you can inherit several interface versus only one class.

MangelMaxime commented:
@alfonsogarciacaro Do you know why Fable standalone would not be able to see the new attribute added to Fable.Core ?

alfonsogarciacaro commented:
IIRC it is because fable-standalone gets Fable.Core and other .NET/F# core libraries from fable-metadata/lib so we need to update the Fable.Core.dll copy there (or skip the tests in fable-standalone for now). I don't remember if the Fable.Core.dll copy in metadata was just compiled normally or using @ncave's method to size down dlls #3531

MangelMaxime commented:
@alfonsogarciacaro That was my suspicion too but because we were having a direct project reference I was not sure.

So in theory, to complete this PR we can just rename the attribute to `[<Pojo>]` or `[<PojoInterface>]` and merge it?

ncave commented:
@MangelMaxime The `Fable.Core.dll` is just bindings (no code), so you can just drop a fresh version in [here](https://github.com/fable-compiler/Fable/tree/main/src/fable-metadata/lib).

alfonsogarciacaro commented:
> That was my suspicion too but because we were having a direct project reference I was not sure.

Also iirc, that is what the `--exclude Fable.Core` option is used for, to ignore compilation of project references:
https://github.com/fable-compiler/Fable/blob/5c95930cf4b778956eade0c5bc702070613ea2b1/src/Fable.Build/Test/Standalone.fs#L23-L28

I updated Fable.Core.dll and renamed the attribute. I left `pojoDefinedByConsArgs` in the internal functions as it is a bit more informative. Name choice is tricky, `Pojo` may be too broad, but `PojoInterface` could be confusing (is it only for F# interfaces?). Maybe `TypescriptPojoInterface`, same as we have `TypescriptTaggedUnion`, but it's also a bit long.

Note: Fable.Core.dll compilation with net9.0 didn't work so at the end I used net6.0. I also had to use Debug mode vs Release mode. I guess it is likely related with .dll trimming, probably fable standalone cannot deal with it.

ncave commented:
@alfonsogarciacaro 
> Fable.Core.dll compilation with net9.0 didn't work

It should work, you just have to turn off the metadata compression using `<CompressMetadata>false</CompressMetadata>`, as current Fable JS standalone cannot handle compressed metadata.

In theory, we could implement partial support for `System.IO.Compression` in JS using [DecompressionStream](https://developer.mozilla.org/en-US/docs/Web/API/DecompressionStream), but I'm not sure it's worth the effort at the moment, so just disable the compression for now. Also, metadata binaries usually get gzipped when downloaded over the internets anyway, so it's fine.

kerams commented:
Alfonso is alive! Great to see you here again.

MangelMaxime commented:
> Name choice is tricky, `Pojo` may be too broad, but `PojoInterface` could be confusing (is it only for F# interfaces?). Maybe `TypescriptPojoInterface`, same as we have `TypescriptTaggedUnion`, but it's also a bit long.

I think `Pojo` is fine because we also specified the attribute target to be used only for classes.

`TypescriptPojoInterface` could also work but this is but raise the same issue as `PojoInterface`, we could also think about `TypescriptInterface` because we could think we want to out this type as a "typescript interface` but the reality is that all the `TypescriptXXX` can also be used when targeting JavaScript.

Pojo is a well known concept in JavaScript, so I think this should be ok.
